src/Dependencies/Collections/Internal/SegmentedArrayHelper.cs

```diff
@@ -14,6 +16,52 @@ internal static class SegmentedArrayHelper
         // Large value types may benefit from a smaller number.
         internal const int IntrosortSizeThreshold = 16;
 
+        /// <summary>
+        /// A combination of <see cref="MethodImplOptions.AggressiveInlining"/> and
+        /// <see cref="F:System.Runtime.CompilerServices.MethodImplOptions.AggressiveOptimization"/>.
+        /// </summary>
+        [SuppressMessage("Documentation", "CA1200:Avoid using cref tags with a prefix", Justification = "The field is not supported in all compilation targets.")]
+        internal const MethodImplOptions FastPathMethodImplOptions = MethodImplOptions.AggressiveInlining | (MethodImplOptions)512;
+
+        [MethodImpl(FastPathMethodImplOptions)]
+        internal static int GetSegmentSize<T>()
+        {
+            if (Unsafe.SizeOf<T>() == Unsafe.SizeOf<object>())
```

Comment:
This approach means our performance will differ between architectures. For instance the `int` will be considered a reference type 32 bit but not in 64 bit. Why aren't we using `typeof(T).IsValueType` here?

Reply:
The only real input to the whole segmenting algorithm is the value of `Unsafe.SizeOf<T>()`. For value types, generic specialization in the JIT means the value for any given value type will be inlined. For reference types, the JIT produces one common assembly implementation, so to avoid overhead we want all reference types to be using the same field reference for this value (both to avoid a lookup table for all the instantiations, and to improve the ability of the JIT to inline the indexer. Value types that happen to have the same size as a reference type are correct on both possible paths, so it doesn't matter which one we use.

`typeof(T).IsValueType` isn't a JIT intrinsic prior to net5.0, so if we try to use it for earlier targets we get something like a 20x regression. `Unsafe.SizeOf<T>()` behaves as an intrinsic (runtime constant) for all targets.